janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/DefaultIndexSelector.java

```diff
@@ -0,0 +1,61 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.query.index;
+
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.condition.Condition;
+import org.janusgraph.graphdb.query.condition.MultiCondition;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.IndexType;
+import org.javatuples.Pair;
+
+import java.util.Set;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class DefaultIndexSelector implements IndexSelectionStrategy {
```

Comment:
This looks more like `Service` because it contains a business logic to interact with different strategies. I would suggest to rename it to `IndexSelectorService` and do not implement `IndexSelectionStrategy`.

Reply:
Why not see this as a Strategy, too? Sure it combines and calls other strategies based on the size of the problem. But from my point of view it serves the same purpose as other implementations of `IndexSelectionStrategy` and acts as a drop-in replacement for other strategies.

Reply:
Good point. I guess it can be seen as a strategy also but from the other hand it just contains logic to choose another strategy based on some logic. I guess if we add another strategy it becomes a dependency of this class also because the logic behind this class is basically to select a strategy.
I see it more as a service but I don't mind leaving it as a strategy. Just give it some more clear name. Maybe something like 'ThresholdBasedIndexSelectionStrategy'.

Reply:
LGTM
I'll see if I can find some spare time today or tomorrow to implement your requested changes